lib/db/db.rb

```diff
@@ -9,26 +9,63 @@ def size
     all.size
   end
 
+  def all
+    reader.nil? ? [] : reader.run
+  end
+
+  def at(index)
+    empty || invalid(index) ? {} : all[index]
+  end
+
   def add(contact)
-    contacts = all
-    contacts << contact
+    contacts = all << format(contact)
     writer.run(contacts) if !writer.nil?
   end
 
-  def all
-    reader.nil? ? [] : reader.run
+  def update(contact)
+    contacts = updated_with(format(contact))
+    writer.run(contacts) if !writer.nil?
   end
 
-  def at(id)
-    all.empty? ? {} : all[id]
+  def search(term)
+    all.select { |contact| contact.any? {|key, val| val.to_s.include?(term)} }
   end
 
-  def search(term)
-    all.select{ |contact| contact.any? {|key, val| val.include?(term)} }
+  def index_of_id(id)
```

Comment:
It is used in the updater, for example, I list the contacts with an index (1, 2, 3...), and the simplest thing seemed to map that to the contacts array index, so that when a user selects a contact I can easily retrieve it using that index. I think the ID is a different concept and should be internal to the DB. Probably it is another wrong decision :(